zipkin-lens/src/slices/tracesSlice.ts

```diff
@@ -0,0 +1,292 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/* eslint-disable no-param-reassign */
+
+import {
+  SerializedError,
+  createAsyncThunk,
+  createSlice,
+} from '@reduxjs/toolkit';
+
+import * as api from '../constants/api';
+import AdjustedTrace from '../models/AdjustedTrace';
+import Span from '../models/Span';
+import TraceSummary from '../models/TraceSummary';
+import { ensureV2TraceData } from '../util/trace';
+
+const {
+  treeCorrectedForClockSkew,
+  traceSummary: buildTraceSummary,
+  traceSummaries: buildTraceSummaries,
+  detailedTraceSummary: buildDetailedTraceSummary,
+} = require('../zipkin');
+
+export const searchTraces = createAsyncThunk(
+  'traces/search',
+  async (params: { [key: string]: string }, thunkApi) => {
+    const ps = new URLSearchParams(params);
+
+    // We need to import RootState in order to give the type to getState.
+    // Importing RootState will result in a cyclic import.
+    // So use any type to avoid this.
+    const { search, traces }: TracesState = (thunkApi.getState() as any).traces;
+    // If the query is the same as the previous query, it will not fetch again.
+    if (search.prevQuery === ps.toString()) {
+      return {
+        traces,
+        traceSummaries: search.traceSummaries,
+        query: ps.toString(),
+      };
+    }
+
+    const resp = await fetch(`${api.TRACES}?${ps.toString()}`);
+    if (!resp.ok) {
+      throw Error(resp.statusText);
+    }
+    const rawTraces: Span[][] = await resp.json();
+
+    const newTraces = rawTraces.reduce(
+      (acc, rawTrace) => {
+        const [{ traceId }] = rawTrace;
+        const skewCorrectedTrace = treeCorrectedForClockSkew(rawTrace);
+        acc[traceId] = {
+          rawTrace,
+          skewCorrectedTrace,
+        };
+        return acc;
+      },
+      {} as {
+        [key: string]: {
+          rawTrace: Span[];
+          skewCorrectedTrace: any;
+        };
+      },
+    );
+
+    const traceSummaries: TraceSummary[] = buildTraceSummaries(
+      ps.get('serviceName'),
+      Object.keys(newTraces).map((traceId) =>
+        buildTraceSummary(newTraces[traceId].skewCorrectedTrace),
+      ),
+    );
+
+    return {
+      traces: newTraces,
+      traceSummaries,
+      query: ps.toString(),
+    };
+  },
+);
+
+export const loadTrace = createAsyncThunk(
+  'traces/load',
+  async (traceId: string, thunkApi) => {
+    // We need to import RootState in order to give the type to getState.
+    // Importing RootState will result in a cyclic import.
+    // So use any type to avoid this.
+    const { traces }: TracesState = (thunkApi.getState() as any).traces;
+
+    if (traces[traceId]) {
+      const { rawTrace, skewCorrectedTrace } = traces[traceId];
+      let { adjustedTrace } = traces[traceId];
+      if (adjustedTrace) {
+        return traces[traceId];
```

Comment:
will the return here have the same type as the result of buildDetailedTraceSummary?

Reply:
Yes.
If this `adjustedTraces` is not `undefined`, this trace has already been calculated by `buildDetailedTraceSummary`, so I avoid recalculating it and return the data as it is.

Reply:
above this `        return traces[traceId];` line, add `// this trace has already been calculated by buildDetailedTraceSummary`?